File: app/src/main/java/org/skyfaced/mvp/service/network/WaifuRepositoryImpl.java
package org.skyfaced.mvp.service.network;

import org.skyfaced.mvp.model.EmptyBody;
import org.skyfaced.mvp.model.ImageDto;
import org.skyfaced.mvp.util.WaifuType;

import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

import io.reactivex.rxjava3.core.Observable;

public class WaifuRepositoryImpl implements WaifuRepository {
    private final WaifuService waifuService;

    public WaifuRepositoryImpl(WaifuService waifuService) {
        this.waifuService = waifuService;
    }

    @Override
    public Observable<ImageDto> waifu(WaifuType type) {
        return waifuService.waifu(type.getName(), type.getCategory())
                .doOnEach(imageDtoNotification -> {
                    if (new Random().nextInt() % 2 == 0) throw new IllegalArgumentException();
                })
                .distinct();
    }

    @Override
    public Observable<List<ImageDto>> waifus(WaifuType type) {
        return waifuService.waifus(type.getName(), type.getCategory(), new EmptyBody())
                .map(it -> it.getFiles().stream().map(ImageDto::new).collect(Collectors.toList()))
                .distinct();
    }
}

File: app/src/main/java/org/skyfaced/mvp/service/network/WaifuRepositoryImpl.kt
package org.skyfaced.mvp.service.network

import io.reactivex.rxjava3.core.Observable
import org.skyfaced.mvp.model.ImageDto
import org.skyfaced.mvp.util.WaifuType
import kotlin.random.Random.Default.nextInt

class WaifuRepositoryImpl(private val waifuService: WaifuService) : WaifuRepository {
    override fun waifu(type: WaifuType): Observable<ImageDto> {
        return waifuService.waifu(type.type, type.category)
            .doOnEach { require(nextInt() % 2 != 0) }
            .distinct()
    }

    override fun waifus(type: WaifuType): Observable<List<ImageDto>> {
        return waifuService.waifus(type.type, type.category)
            .map { it.files.map(::ImageDto) }
            .distinct()
    }
}